src/crimson/osd/replicated_recovery_backend.cc

```diff
@@ -652,29 +641,19 @@ seastar::future<bool> ReplicatedRecoveryBackend::_handle_pull_response(
   if (pi.recovery_info.version == eversion_t())
     pi.recovery_info.version = pop.version;
 
-  bool first = pi.recovery_progress.first;
-
-  return [this, &pi, first, &recovery_waiter, &pop] {
-    if (first) {
-      return pg.get_or_load_head_obc(pi.recovery_info.soid).safe_then(
-	[&pi, &recovery_waiter, &pop](auto p) {
-	auto& [obc, existed] = p;
-	pi.obc = obc;
-	recovery_waiter.obc = obc;
-	obc->obs.oi.decode(pop.attrset[OI_ATTR]);
-	pi.recovery_info.oi = obc->obs.oi;
-	return seastar::make_ready_future<>();
-      }, crimson::osd::PG::load_obc_ertr::all_same_way(
-	  [this, &pi](const std::error_code& e) {
-	  auto [obc, existed] = shard_services.obc_registry.get_cached_obc(
-				    pi.recovery_info.soid);
-	  pi.obc = obc;
-	  return seastar::make_ready_future<>();
-	})
-      );
-    }
-    return seastar::make_ready_future<>();
-  }().then([this, first, &pi, &pop, t, response]() mutable {
+  auto prepare_waiter = seastar::make_ready_future<>();
+  if (pi.recovery_progress.first) {
+    prepare_waiter = pg.with_head_obc<RWState::RWNONE>(
```

Comment:
i assume that we cannot serve client request and we cannot perform recovery on the same object when we are already recovering it. so i am not acquiring the exclusive lock here. and i also removed `obc->put_lock_type(RWState::RWEXCL)` in the `PGRecovery::on_local_recover()` call.